### PR TITLE
ARROW-2462: [C++] Fix Segfault in UnpackBinaryDictionary

### PR DESCRIPTION
**Pull request by zeroshade**
Discovered this through using pyarrow and dealing with RecordBatch Streams and parquet. The issue can be replicated as follows:

```python
import pyarrow as pa
import pyarrow.parquet as pq

# create record batch with 1 dictionary column
indices = pa.array([1,0,1,1,0])
dictionary = pa.array(['Foo', 'Bar'])
dict_array = pa.DictionaryArray.from_arrays(indices, dictionary)
rb = pa.RecordBatch.from_arrays( [ dict_array ], [ 'd0' ] )

# write out using RecordBatchStreamWriter
sink = pa.BufferOutputStream()
writer = pa.RecordBatchStreamWriter(sink, rb.schema)
writer.write_batch(rb)
writer.close()
buf = sink.get_result()

# read in and try to write parquet table
reader = pa.open_stream(buf)
tbl = reader.read_all()
pq.write_table(tbl, 'dict_table.parquet') # SEGFAULTS
```

When writing record batch streams, if there are no nulls in an array, Arrow will put a placeholder nullptr instead of putting the full bitmap of 1s, when deserializing that stream, the bitmap for the nulls isn't populated and is left to being a nullptr. When attempting to write this table via pyarrow.parquet, you end up [here](https://github.com/apache/parquet-cpp/blob/master/src/parquet/arrow/writer.cc#L963) in the parquet writer code which attempts to Cast the dictionary to a non-dictionary representation. Since the null count isn't checked before creating a BitmapReader, the BitmapReader is constructed with a nullptr for the bitmap_data, but a non-zero length which then segfaults in the constructor [here](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bit-util.h#L415) because `bitmap` is null. 

So a simple check of the null count before constructing the BitmapReader avoids the segfault.

### CODE REVIEW

**Comment by xhochy** [comment]
Change looks good but has formatting issues. @zeroshade can you run `make format` and commit the changes?

**Comment by zeroshade** [comment]
@xhochy Added the changes from `make format` :) I was surprised that it updated files outside of the original change.

**Comment by zeroshade** [comment]
@xhochy Ok, now I'm confused, I ran `make check-format` myself before pushing the commits and it passed. Is it because i squashed the commits that it failed on here?

**Comment by pitrou** [comment]
@zeroshade Perhaps you are not using the right clang version. You need `clang-format-5.0` installed for `make format` to produce output exactly compatible with the CI expectations.

**Comment by zeroshade** [comment]
@pitrou That seems to have been the issue, I downgraded to clang-format 5.0 on my machine instead of the v6 i was using, and it makes sense. So hopefully it should pass this time. Thanks for the suggestion there.

**Comment by zeroshade** [comment]
The travis CI build failed due to a nodejs error:

```
npm ERR! code EINTEGRITY
npm ERR! sha512-bqB1yS6o9TNA9ZC/MJxM0FZzPnZdtHj0xWK/IZ5khzVqdpGul/R/EIiHRgFXlwTD7PSIaYVnGKq1QgMCu2mnqw== integrity checksum failed when using sha512: wanted sha512-bqB1yS6o9TNA9ZC/MJxM0FZzPnZdtHj0xWK/IZ5khzVqdpGul/R/EIiHRgFXlwTD7PSIaYVnGKq1QgMCu2mnqw== but got sha512-0359nisgdDWWKoe3o8//wD4beo1dN9RDCKpKToPk5grFGRWxMPGg+bsdP651X+wrrtxllqw/oPARk10vvMiYXg==. (359772 bytes)
```

I'm not familiar enough with nodejs here is there a checksum I need to update due to the c++ changes? or is this failure unrelated to my change?

**Comment by zeroshade** [comment]
@pitrou huzzah! Thanks much! yay! it succeeded! :smiley:

**Review by xhochy — COMMENTED** [review]
Looks good, just adding a needs work here, so I know that something is missing.

@zeroshade if you need help for the test, feel free to reach out.

**Review by xhochy — CHANGES_REQUESTED** [review]
Looks good, just adding a needs work here, so I know that something is missing.

@zeroshade if you need help for the test, feel free to reach out.

**Comment by zeroshade** [comment]
@xhochy I'm actually in the middle of packing and moving into a house I just bought haha so I don't know when I'll have time to get to write the test.

So if you could write the test that'd be awesome, otherwise I should be able to get to it in the next couple days. Thanks!

File: cpp/src/arrow/compute/compute-test.cc [hunk 1]
```diff
@@ -790,6 +790,30 @@ TYPED_TEST(TestDictionaryCast, Basic) {
...
+                             std::make_shared<DictionaryArray>(dict_type, c2)};
+  auto dict_carr = std::make_shared<ChunkedArray>(dict_arrays);
+
+  Datum cast_input(dict_carr);
+  Datum cast_output;
+  // Ensure that casting works even when the null_bitmap_data array is a nullptr
+  ASSERT_OK(Cast(&this->ctx_, cast_input,
+                 static_cast<DictionaryType&>(*dict_type).dictionary()->type(),
+                 CastOptions(), &cast_output));
+  ASSERT_EQ(Datum::CHUNKED_ARRAY, cast_output.kind());
```
Comment (pitrou) [inline cpp/src/arrow/compute/compute-test.cc]: Shouldn't the casted values be checked too?

**Comment by zeroshade** [comment]
@pitrou Done :smile:

**Review by pitrou — APPROVED** [review]
LGTM.

**Comment by zeroshade** [comment]
Hey @pitrou why was this closed instead of merged?

**Comment by pitrou** [comment]
No problem :)

**Pull request closed without merging**